### PR TITLE
[FLINK-34613][cdc] Support recover from a specific savepoint file

### PR DESCRIPTION
**Pull request by lvyanquan**
This closes https://github.com/ververica/flink-cdc-connectors/issues/2940.
Use `fromSavepoint`  to be consistent with [Flink cli](https://nightlies.apache.org/flink/flink-docs-master/docs/deployment/cli/#starting-a-job-from-a-savepoint).  Some codes are inspired from [CliFrontendParser](https://github.com/apache/flink/blob/6e78eb18524ead3abd60da0ca41751b45e0e2482/flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java#L681-L700).

### CODE REVIEW

**Comment by lvyanquan** [comment]
@PatrickRen  PTAL.

**Comment by Jazzylol** [comment]
@lvyanquan  您好，我使用您的PR打了包，结果报错，错误如下  Exception in thread "main" com.ververica.cdc.shaded.com.apache.commons.cli.UnrecognizedOptionException: Unrecognized option: --fromSavepoint
        at com.ververica.cdc.shaded.com.apache.commons.cli.DefaultParser.handleUnknownToken(DefaultParser.java:561)
        at com.ververica.cdc.shaded.com.apache.commons.cli.DefaultParser.handleLongOptionWithoutEqual(DefaultParser.java:383)
        at com.ververica.cdc.shaded.com.apache.commons.cli.DefaultParser.handleLongOption(DefaultParser.java:334)
        at com.ververica.cdc.shaded.com.apache.commons.cli.DefaultParser.handleToken(DefaultParser.java:540)
        at com.ververica.cdc.shaded.com.apache.commons.cli.DefaultParser.parse(DefaultParser.java:712)
        at com.ververica.cdc.shaded.com.apache.commons.cli.DefaultParser.parse(DefaultParser.java:679)
        at com.ververica.cdc.shaded.com.apache.commons.cli.DefaultParser.parse(DefaultParser.java:660)
        at com.ververica.cdc.cli.CliFrontend.main(CliFrontend.java:50)

执行的命令 bash /opt/flink-cdc-3.0.0/bin/flink-cdc.sh /opt/flink-cdc-3.0.0/conf/mysql-to-starrocks.yaml --fromSavepoint file:/opt/flink-storage/savepoint/savepoint-f414a3-f2540c5684d0

请问是有什么需要注意的地方吗？

**Comment by lvyanquan** [comment]
@Jazzylol Have you updated `lib/flink-cdc-dist-3.0-SNAPSHOT.jar`? I have tested again and run correctly.

File: flink-cdc-cli/src/main/java/com/ververica/cdc/cli/CliFrontendOptions.java [hunk 1]
```diff
@@ -52,12 +52,35 @@ public class CliFrontendOptions {
...
+    public static final Option SAVEPOINT_PATH_OPTION =
+            Option.builder("s")
+                    .longOpt("fromSavepoint")
+                    .hasArg(true)
+                    .desc(
+                            "Path to a savepoint to restore the job from (for example hdfs:///flink/savepoint-1537")
+                    .build();
+
+    public static final Option SAVEPOINT_RESTORE_MODE =
+            Option.builder("rm")
```
Comment (leonardBang) [inline flink-cdc-cli/src/main/java/com/ververica/cdc/cli/CliFrontendOptions.java:64]: `rm` is a dangerous abbreviation in CS which usually means `remove`, please avoid use this abbreviation

File: flink-cdc-cli/src/main/java/com/ververica/cdc/cli/CliFrontendOptions.java [hunk 2]
```diff
@@ -52,12 +52,35 @@ public class CliFrontendOptions {
...
+                    .longOpt("restoreMode")
```
Comment (leonardBang) [inline flink-cdc-cli/src/main/java/com/ververica/cdc/cli/CliFrontendOptions.java:65]:
```suggestion
                    .longOpt("savepointRestoreMode")
```

File: flink-cdc-cli/src/test/java/com/ververica/cdc/cli/CliFrontendTest.java [hunk 3]
```diff
@@ -137,6 +156,16 @@ private String globalPipelineConfig() throws Exception {
                     + "                               CDC pipelines\n"
                     + "    -h,--help                  Display help message\n"
                     + "       --jar <arg>             JARs to be submitted together with the pipeline\n"
+                    + "    -rm,--restoreMode <arg>    Defines how should we restore from the given\n"
```
Comment (leonardBang) [inline flink-cdc-cli/src/test/java/com/ververica/cdc/cli/CliFrontendTest.java:159]: Could we consider other savepoint settings like `allowNonRestoredState`?

**Review by leonardBang — COMMENTED** [review]
Thanks @lvyanquan for the contribution, I left some comments

**Comment by lvyanquan** [comment]
Addressed it.

**Review by leonardBang — APPROVED** [review]
Thanks @lvyanquan for the update, LGTM

**Comment by PRNDA** [comment]
Work like a charm.

File: flink-cdc-cli/src/main/java/org/apache/flink/cdc/cli/CliFrontendOptions.java [hunk 4]
```diff
@@ -53,12 +53,46 @@ public class CliFrontendOptions {
                     .desc("Use Flink MiniCluster to run the pipeline")
                     .build();
 
+    public static final Option SAVEPOINT_PATH_OPTION =
+            Option.builder("s")
+                    .longOpt("fromSavepoint")
```
Comment (PatrickRen) [inline flink-cdc-cli/src/main/java/org/apache/flink/cdc/cli/CliFrontendOptions.java:58]: What about using `from-savepoint` to align with other options? Same issue for options below

**Review by PatrickRen — COMMENTED** [review]
@lvyanquan Thanks for the pull request! I left a comment.

File: flink-cdc-cli/src/main/java/org/apache/flink/cdc/cli/CliFrontendOptions.java (hunk 4 above)
Comment (lvyanquan) [reply flink-cdc-cli/src/main/java/org/apache/flink/cdc/cli/CliFrontendOptions.java]: Addressed it.

**Review by PatrickRen — APPROVED** [review]
@lvyanquan Thanks for the update! LGTM.

**Comment by monologuist** [comment]
@lvyanquan hi，请问您这个PR，所预期的从savepoint重启Flink CDC Pipeline作业的规范命令是这样吗？

```
在${FLINK_CDC_HOME}目录下执行：
$ ./bin/flink-cdc.sh mysql-to-doris.yaml -s /tmp/flink-savepoints/savepoint-cca7bc-bb1e257f0dab
```

因为我没有看到明确的重启命令说明*，所以对此怀有疑惑。并且我认为应当明确指出。

*您引用的链接中，我看到的仍是针对jar作业的重启方式。
$ ./bin/flink run \
      --detached \ 
      --from-savepoint /tmp/flink-savepoints/savepoint-cca7bc-bb1e257f0dab \
      ./FlinkCDCExample.jar

**Comment by monologuist** [comment]
> @lvyanquan hi，请问您这个PR，所预期的从savepoint重启Flink CDC Pipeline作业的规范命令是这样吗？
> 
> ```
> 在${FLINK_CDC_HOME}目录下执行：
> $ ./bin/flink-cdc.sh mysql-to-doris.yaml -s /tmp/flink-savepoints/savepoint-cca7bc-bb1e257f0dab
> ```
> 
> 因为我没有看到明确的重启命令说明*，所以对此怀有疑惑。并且我认为应当明确指出。
> 
> *您引用的链接中，我看到的仍是针对jar作业的重启方式。 $ ./bin/flink run --detached \ --from-savepoint /tmp/[flink-savepoints](https://issues.apache.org/jira/browse/FLINK-savepoints)/savepoint-cca7bc-bb1e257f0dab ./FlinkCDCExample.jar

经多次测试，我确认这个命令可行。

**Pull request merged**